pkg/kv/kvserver/replica_raft.go

```diff
@@ -2615,9 +2615,14 @@ func (r *Replica) campaignLocked(ctx context.Context) {
 // TODO(nvanbenschoten): this is the remaining logic which needs work in order
 // to complete #125254. See the comment in raft.go about how even a local
 // fortification check is not enough to make MsgTimeoutNow safe.
-func (r *Replica) forceCampaignLocked(ctx context.Context) {
+func (r *Replica) forceCampaignLocked(ctx context.Context, lead raftpb.PeerID) {
 	log.VEventf(ctx, 3, "force campaigning")
-	msg := raftpb.Message{To: raftpb.PeerID(r.replicaID), Type: raftpb.MsgTimeoutNow}
+	msg := raftpb.Message{
+		To: raftpb.PeerID(r.replicaID),
+		// We pretend that the message was sent from the leader.
+		From: lead,
+		Type: raftpb.MsgTimeoutNow,
+	}
```

Comment:
> Is this related to the change in this commit, or a drive-by? I see that From is used in the handler for logging, is that why?

It's related to the change and to our thread above. The use of `MsgTimeoutNow` like this is a hack and is hardly safe, it's only maybe ok because of when this code is run (see `shouldCampaignAfterConfChange`). I'd like to get rid of this, but in the meantime, I'm ok making it even more clear that cockroach is lying to raft about this message coming from the leader, which in some ways it is because the leader is the one committing the entry which demotes itself to a learner.

> Also, is it safe/canonical to communicate a leader without communicating the term? I imagine if a follower receives a MsgTimeoutNow for a long gone term, it should ignore.

Good point. I added the `Term`, along with a better comment.